lib/hal/xbox.c

```diff
@@ -9,51 +9,78 @@
+        if (LaunchDataPage == NULL) {
+            return -1;
+        }
+        memset(LaunchDataPage, 0, LaunchDataPageSize);
+        LaunchDataPage->Header.dwLaunchDataType = LDT_FROM_DASHBOARD;
+    }
+
+    *launchDataType = LaunchDataPage->Header.dwLaunchDataType;
+    *launchData = LaunchDataPage->LaunchData;
+    return 0;
 }
 
-/**
- * Launches an XBE.  Examples of xbePath might be:
- *   c:\\blah.xbe
- *   c:/foo/bar.xbe
- * If the XBE is able to be launched, this method will
- * not return.  If there is a problem, then it return.
- */
-void XLaunchXBE(char *xbePath)
+void XLaunchXBE(const char *xbePath)
 {
-#if 0
-	struct stat statbuf;
-	if (stat(xbePath, &statbuf) < 0)
-		return;
-#endif
+    XLaunchXBEEx(xbePath, NULL);
+}
 
-    if (LaunchDataPage == NULL)
+void XLaunchXBEEx(const char *xbePath, const void *launchData)
+{
+    if (LaunchDataPage == NULL) {
         LaunchDataPage = MmAllocateContiguousMemory(0x1000);
+        if (LaunchDataPage == NULL) {
+            return;
+        }
+    }
+
+    // For ease of debugging.
+    PLAUNCH_DATA_PAGE launchDataPage = LaunchDataPage;
+
+    MmPersistContiguousMemory(launchDataPage, LaunchDataPageSize, TRUE);
+    memset((void*)launchDataPage, 0, LaunchDataPageSize);
 
-    if (LaunchDataPage == NULL)
-		return;
+    launchDataPage->Header.dwLaunchDataType = LDT_TITLE;
+    launchDataPage->Header.dwTitleId = CURRENT_XBE_HEADER->CertificateHeader->TitleID;
+    launchDataPage->Header.dwFlags = 0x0000;
 
-    MmPersistContiguousMemory(LaunchDataPage, 0x1000, TRUE);
+    if (!xbePath) {
+        launchDataPage->Header.dwLaunchDataType = LDT_LAUNCH_DASHBOARD;
+    } else {
+        XConvertDOSFilenameToXBOX(xbePath, launchDataPage->Header.szLaunchPath);
 
-	memset((void*)LaunchDataPage, 0, 0x1000);
+        // one last thing... xbePath now looks like:
+        //   \Device\Harddisk0\Partition2\blah\doom.xbe
+        // but it has to look like:
+        //   \Device\Harddisk0\Partition2\blah;doom.xbe
+        char *lastSlash = strrchr(launchDataPage->Header.szLaunchPath, '\\');
+        if (!lastSlash) {
+            // if we couldn't find a trailing slash, the conversion to
+            // the xbox path mustn't have worked, so we will return
+            return;
+        }
 
-	LaunchDataPage->Header.dwLaunchDataType = 0xFFFFFFFF;
-	LaunchDataPage->Header.dwTitleId = 0;
-	XConvertDOSFilenameToXBOX(xbePath, LaunchDataPage->Header.szLaunchPath);
+        *lastSlash = ';';
+    }
 
-	// one last thing... xbePath now looks like:
-	//   \Device\Harddisk0\Partition2\blah\doom.xbe
-	// but it has to look like:
-	//   \Device\Harddisk0\Partition2\blah;doom.xbe
-	char *lastSlash = strrchr(LaunchDataPage->Header.szLaunchPath, '\\');
-	if (lastSlash != NULL)
-	{
-		*lastSlash = ';';
-		HalReturnToFirmware(HalQuickRebootRoutine);
-	}
+    if (launchData) {
+        memcpy(launchDataPage->LaunchData, launchData, sizeof(launchDataPage->LaunchData));
+    }
```

Comment:
If `launchData` is `NULL` the `LaunchDataPage` will still be allocated and point at zero-bytes; is this intended?